File: 2017/day12.go
package main
import (
        "fmt"
        "bufio"
        "os"
	"strings"
	"strconv"
)

func main() {
	file, _ := os.Open("day12.input")
        defer file.Close()
        scanner := bufio.NewScanner(file)
	gates := make(map[int] []int)
	var gatesList []int
	gatesToDest := make(map[int] byte) //0 is unknown, 1 is confirmed gate to zero, 2 is visited and unknown
	dest := 0
	sumA := 0
	countB := 0

	for scanner.Scan() {
		line := scanner.Text()
		input := strings.Split(line, " <-> ")
		var destinations []int
		source, _ := strconv.Atoi(input[0])
		for _,i := range strings.Split(input[1], ", ") {
			val, _ := strconv.Atoi(i)
			destinations = append(destinations,val)
		}
		gates[source] = destinations
		gatesToDest[source] = 0
		gatesList = append(gatesList, source)
	}

	for dest != 2000000 {
		nextGates := make(map[int] []int)
		nextGatesToDest := make(map[int] byte)
		gatesToDest[0] = 1
		var callstack []int
		solve(dest, gatesList, gates, gatesToDest, dest, callstack)
		nextDest := 2000000
		for k, v := range gatesToDest {
			if v == 1 && dest == 0 {
				sumA++
			} else if v == 2 {
				if k < nextDest {
					nextDest = k
				}
				nextGates[k] = gates[k]
				nextGatesToDest[k] = 0
			}
		}
		dest = nextDest
		gates = nextGates
		gatesToDest = nextGatesToDest
		countB++
	}

	fmt.Printf("Answer part A: %v\n", sumA)
	fmt.Printf("Answer part B: %v\n", countB)
}

func solve(dest int, gatesList []int, gatesDict map[int][]int, gatesToDest map[int] byte, start int, callstack []int) {
	for i := start; i < start + len(gatesList); i++ {
		gate := gatesList[i % len(gatesList)]
		destinations := gatesDict[gate]
		if gatesToDest[gate] != 1 {
			for _, destination := range destinations {
				if destination == dest || gatesToDest[destination] == 1 {
					callstack = append(callstack, gate)
					callstack = append(callstack, destinations...)
					for _, g := range callstack {
						gatesToDest[g] = 1
					}
					callstack = nil
					break
				} else if gatesToDest[gate] == 0 {
					gatesToDest[gate] = 2
					solve(dest, gatesList, gatesDict, gatesToDest, destination, append(callstack, gate))
				} else if gatesToDest[gate] == 2 {
					callstack = nil
				}
			}
		}
	}
}

File: 2020/python/day12.py
input = [line.strip() for line in open("../input/day12.input").readlines()]
DD = {"N":(0,-1), "W":(-1,0), "S":(0,1), "E":(1,0)}
course, Pa, W, Pb = "E", (0, 0), (10, -1), (0,0)
for line in input:
  d, count = line[0], int(line[1:])
  if d in "NWSE":
    dx,dy = count * DD[d][0], count * DD[d][1]
    Pa = (Pa[0]+dx, Pa[1]+dy)
    W = (W[0]+dx, W[1]+dy)
  elif d == "F":
    dx,dy = count * DD[course][0], count * DD[course][1]
    Pa = (Pa[0]+dx, Pa[1]+dy)
    Pb = (Pb[0]+count * W[0], Pb[1]+count * W[1])
  else:
    for _ in range(count//90): W = (W[1] * -1, W[0]) if d == "R" else (W[1], W[0] * -1)
    course = "NWSE"[("NWSE".index(course) - count//90) % 4] if d == "R" else "NWSE"[("NWSE".index(course) + count//90) % 4]
print("Answer A:", abs(Pa[0]) + abs(Pa[1]))
print("Answer B:", abs(Pb[0]) + abs(Pb[1]))
